lib/linguist/heuristics.rb

```diff
@@ -29,33 +32,52 @@ def self.call(blob, candidates)
       [] # No heuristics matched
     end
 
-    # Internal: Define a new heuristic.
-    #
-    # exts_and_langs - String names of file extensions and languages to
-    #                  disambiguate.
-    # heuristic - Block which takes data as an argument and returns a Language or nil.
-    #
-    # Examples
-    #
-    #     disambiguate ".pm" do |data|
-    #       if data.include?("use strict")
-    #         Language["Perl"]
-    #       elsif /^[^#]+:-/.match(data)
-    #         Language["Prolog"]
-    #       end
-    #     end
+    # Internal: Load heuristics from 'heuristics.yml'.
+    def self.load()
+      if @heuristics.any?
+        return
+      end
+
+      data = YAML.load_file(File.expand_path("../heuristics.yml", __FILE__))
+      named_patterns = data['named_patterns'].map { |k,v| [k, self.to_regex(v)] }.to_h
+
+      data['disambiguations'].each do |disambiguation|
+        exts = disambiguation['extensions']
+        rules = disambiguation['rules']
+        rules.map! do |rule|
+          if !rule['pattern'].nil?
+            rule['pattern'] = self.to_regex(rule['pattern'])
+          elsif !rule['negative_pattern'].nil?
+            pat = self.to_regex(rule['negative_pattern'])
+            rule['pattern'] = NegativePattern.new(pat)
+          elsif !rule['named_pattern'].nil?
+            rule['pattern'] = named_patterns[rule['named_pattern']]
+          end
+          rule
+        end
+        @heuristics << new(exts, rules)
+      end
+    end
+
+    # Internal: Converts a string or array of strings to regexp
     #
-    def self.disambiguate(*exts_and_langs, &heuristic)
-      @heuristics << new(exts_and_langs, &heuristic)
+    # str: string or array of strings. If it is an array of strings,
+    #      Regexp.union will be used.
+    def self.to_regex(str)
+      if str.kind_of?(Array)
+        Regexp.union(str.map { |s| Regexp.new(s) })
+      else
+        Regexp.new(str)
+      end
     end
 
     # Internal: Array of defined heuristics
     @heuristics = []
 
     # Internal
-    def initialize(exts_and_langs, &heuristic)
+    def initialize(exts_and_langs, rules)
       @exts_and_langs, @candidates = exts_and_langs.partition {|e| e =~ /\A\./}
```

Comment:
It looks like the YAML file doesn't support candidates (which is fine since we don't use that anymore). Thus, the above line can be simplified, as well as the additional logic in `matches?` to handle candidates.

Reply:
I'll perform this change next time I get to work on this.